test/core.jl

```diff
@@ -3499,3 +3499,10 @@ let
     finalize(obj)
     @test finalized == 1
 end
+
+# non-terminating inference, issue #14009
+type A14009{T}
+end
+A14009{T}(a::T) = A14009{T}()
+f14009(a) = ? ? f14009(A14009(a)) : a
```

Comment:
sure, I happen to really like this but I think I'm alone on this one. I somewhat remember convincing Jeff that it made question-mark-as-a-global a feature but maybe he was just being nice.